book/src/tooling/coap.md

```diff
@@ -0,0 +1,265 @@
+and offers features exceeding the classical REST set such as [observation].
+
+**RIOT-rs supports** the use of CoAP
+for implementing clients, servers or both in a single device.
+As there is more complexity in a CoAP client than a CoAP server,
+the most minimal use is to just have a CoAP server on the embedded system.
+Having both components in use is also a common choice.
+
+Also, as part of our mission for strong security,
+we use encrypted CoAP traffic by default as explained below.
+
+[CoAP]: https://coap.space/
+[over UDP]: https://datatracker.ietf.org/doc/html/rfc7252
+[over TCP and WebSockets]: https://datatracker.ietf.org/doc/html/rfc8323
+[over SMS and NB-IoT]: https://www.omaspecworks.org/wp-content/uploads/2018/10/Whitepaper-11.1.18.pdf
+[observation]: https://datatracker.ietf.org/doc/html/rfc7641
+
+## Usage: Server side
+
+An example of a CoAP server is [provided as `examples/coap`], see [its `run()` function] for the practical steps.
+
+A CoAP server is created by assembling several **resource handlers on dedicated paths**:
+There might be a path `/s/0` representing a particular sensor,
+and a path `/fwup` for interacting with firmware updates.
+The CoAP implementation can put additional resources at well-known locations,
+eg. `/.well-known/core` for discovery or `/.well-known/edhoc` for establishing secure connections.
+
+The handler needs to concern itself with security aspects of the request content
+(eg. file format parsers should treat incoming data as possibly malformed),
+but the decision whether or not a request is allowed is delegated to an [access policy](#server-access-policy).
+
+[provided as `examples/coap`]: https://github.com/future-proof-iot/RIOT-rs/tree/main/examples/coap
+[its `run()` function]: https://github.com/future-proof-iot/RIOT-rs/blob/2b76e560394884d3c8f7eaae51beefd59a316d7b/examples/coap/src/main.rs#L70
+
+
+## Usage: Client side
+
+The example [provided as `examples/coap`] also contains client steps.
+
+A program that triggers a CoAP request provides[^whatsinarequest] some components to the CoAP stack before phrasing the actual request:
+
+* A **URL describing the resource**, eg. `coap://coap.summit.riot-os.org/agenda` or `coap+tcp://[2001:db8::1]/.well-known/core`.
+
+  Note that while the address is printed here as a text URL
+  (and may even be entered as such in code),
+  it is kept in memory and sent as binary components.
+
+* Optionally, directions regarding how to reach or find the host.
+
+  This is not needed when there are IP literals involved, or DNS is available to resolve names.
+  Directions may be available if the application discovered a usable proxy,
+  or when it is desired to use opt-in discovery mechanisms such as multicast.
+
+* A policy reference on how to authenticate the server, and which identity to present.
+  This is optional if there is a global policy,
+  or if there is an implied security mechanism for the origin.
```

Comment:
"origin" here means URL?

Reply:
It means something [more precise](https://datatracker.ietf.org/doc/html/rfc6454), but explaining the differences would be excessive, so I just put URL instead.